src/test/kotlin/com/github/davenury/ucac/utils/TestApplicationSet.kt

```diff
@@ -8,56 +8,54 @@ import kotlin.random.Random
 class TestApplicationSet(
     numberOfPeersets: Int,
     numberOfPeersInPeersets: List<Int>,
-    actions: Map<Int, Map<TestAddon, suspend (ProtocolTestInformation) -> Unit>> = emptyMap(),
+    actions: Map<Int, Map<TestAddon, AdditionalAction>> = emptyMap(),
     eventListeners: Map<Int, List<EventListener>> = emptyMap(),
-    configOverrides: Map<String, Any> = emptyMap()
+    configOverrides: Map<Int, Map<String, Any>> = emptyMap(),
+    appsToExclude: List<Int> = emptyList()
 ) {
 
-    private val apps: MutableList<MutableList<Application>> = mutableListOf()
+    private var apps: MutableList<MutableList<Application>> = mutableListOf()
     private val peers: List<List<String>>
 
     init {
-
         val ratisConfigOverrides = mapOf(
             "raft.server.addresses" to List(numberOfPeersets) {
                 List(numberOfPeersInPeersets[it]) { "localhost:${Random.nextInt(5000, 20000) + 11124}" }
             },
             "raft.clusterGroupIds" to List(numberOfPeersets) { UUID.randomUUID() }
         )
 
-        // applications creation
-        repeat(numberOfPeersets) { peersetId ->
-            val peersetApps = mutableListOf<Application>()
-            repeat(numberOfPeersInPeersets[peersetId]) { peerId ->
-                val absoluteIndex = apps.flatten().size + peerId + 1
-                peersetApps.add(
-                    createApplication(
-                        arrayOf("${peerId + 1}", "${peersetId + 1}"),
-                        actions[absoluteIndex] ?: emptyMap(),
-                        eventListeners[absoluteIndex] ?: emptyList(),
-                        ratisConfigOverrides + configOverrides,
-                        TestApplicationMode(peerId + 1, peersetId + 1)
-                    )
+        var currentApp = 0
+        apps = MutableList(numberOfPeersets) { peersetId ->
+            MutableList(numberOfPeersInPeersets[peersetId]) { peerId ->
+                currentApp++
+                createApplication(
+                    arrayOf("${peerId + 1}", "${peersetId + 1}"),
+                    actions[currentApp] ?: emptyMap(),
+                    eventListeners[currentApp] ?: emptyList(),
+                    ratisConfigOverrides + (configOverrides[peerId] ?: emptyMap()),
+                    TestApplicationMode(peerId + 1, peersetId + 1)
                 )
             }
-            apps.add(peersetApps)
         }
 
         // start and address discovery
-        apps.flatten().forEach { app -> app.startNonblocking() }
+        apps.flatten().forEachIndexed { index, app -> if (index + 1 !in appsToExclude) app.startNonblocking() }
         peers =
             apps.flatten()
                 .asSequence()
-                .withIndex()
-                .groupBy{ it.value.getPeersetId() }
+                .mapIndexed {index, it -> Pair(it, if (index + 1 in appsToExclude) "localhost:0" else "localhost:${it.getBoundPort()}") }

```

Comment:
What do you think about this change?
```suggestion
                .mapIndexed {index, it -> 
                val address = if (index + 1 in appsToExclude) "localhost:0" else "localhost:${it.getBoundPort()}"
                Pair(it, address) 
                }

```